crates/federated-server/src/server/mcp/server.rs

```diff
@@ -0,0 +1,672 @@
+        if the return type is an object, a union or an interface.
+
+        If it's an object, you have to select at least one field from the type.
+
+        If it's a union, you can select any fields from any of the possible types. Remember to use fragment spreads,
+        if needed.
+
+        If it's an interface, you can select any fields from any of the possible types, or with fields
+        from the interface itself. Remember to use fragment spreads, if needed.
+    "#};
+
+    // Add simple string selection parameter
+    properties.insert(
+        "__selection".to_string(),
+        json!({
+            "type": "string",
+            "description": description,
+        }),
+    );
+
+    required_arguments.push("__selection".to_string());
+
+    let parameters = json!({
+        "type": "object",
+        "properties": properties,
+        "required": required_arguments,
+    });
+
+    let field_description = field.description().unwrap_or("");
+
+    let description = formatdoc! {r#"
+        {field_description}
+
+        This {tool_type} returns a {type} named {type_name}. It is {type_description}.
+        Provide a GraphQL selection set for the query (e.g., '{{ id name }}').
+
+        You must determine the fields of the type by calling the `introspect-type` tool first in
+        this MCP server. It will return the needed information for you to build the selection.
+
+        Do NOT call this {tool_type} before running introspection and knowing exactly what fields you can select.
+    "#};
+
+    tools.push(Tool::new(
+        format!("{tool_type}/{}", field.name()),
+        description.trim_start().to_string(),
+        parameters.as_object().unwrap().clone(),
+    ));
+}
+
+#[must_use]
+fn create_input_properties(input: InputObjectDefinition<'_>) -> serde_json::Value {
+    let mut properties = Map::new();
+    let mut required = Vec::new();
+
+    for field in input.input_fields() {
+        if field.ty().wrapping.is_required() {
+            required.push(field.name());
+        }
+
+        add_argument_to_properties(
+            &mut properties,
+            field.ty().definition(),
+            field.name(),
+            field.description().unwrap_or_default(),
+        );
+    }
+
+    json!({
+        "type": "object",
+        "properties": properties,
+        "required": required,
+        "description": input.description().unwrap_or_default(),
+    })
+}
+
+fn add_argument_to_properties(
+    properties: &mut Map<String, Value>,
+    definition: Definition<'_>,
+    name: &str,
+    description: &str,
+) {
+    match definition {
+        Definition::Scalar(scalar) => {
+            let r#type = match scalar.ty {
+                ScalarType::String => "string",
+                ScalarType::Float => "number",
+                ScalarType::Int => "integer",
+                ScalarType::BigInt => "integer",
+                ScalarType::Boolean => "boolean",
+                ScalarType::Unknown => "string",
```

Comment:
don't know if you have that choice, but `Unknown` is basically arbitrary json rather than a string